Modified calculate_K_trans to compute tau between a list of points within input_data.

damianjilk2 opened the pull request:
Current version consists of an updated tau calculation where it now considers a list of positions acquired from input_data.json.
Future work includes:
- Add tests/validation to ensure tau is being calculated correctly (or at least as expected)
- Implement default position calculation for voxels set to center
- Handle scenarios where the user inputs multiple values for one voxel
     - Average the tau values to find streaming operator and then flux
     - Perhaps consider other averages as well (such as averaging the streaming operator or increasing matrices and averaging the flux)
- Implement input validation and edge handling

damianjilk2 committed:
Modified calculate_K_trans to compute tau between a list of points within input_data. Issue #22

damianjilk2 committed:
Modified calculate_K_trans to compute tau between a list of points within input_data. Issue #22

damianjilk2 commented:
Too much time has been spent on configuring an implementation for the user to input positions. The next version will simply have the user input a dictionary depicting the voxel index and the physical position they wish to calculate. Future versions should also allow for multiple points per voxel.

damianjilk2 commented:
An issue was brought up during the neutronics meeting noting a positive value for the streaming operator during intra-voxel calculations. Here is the referenced conversation: 

Some thoughts:
We (I?) keep conceiving of this 1-D problem as being finite in the other dimensions.  This isn’t true for a real 1-D problem.  In a real 1-D problem, those other dimensions are infinite and I think are the origins of the exponential integrals in the L&M “Slab Geometry” derivation
It’s good for you to be able to explain why this is true.
If you are going to do the 1-D case properly as a slab geometry, you’ll need to use Eqn 5-30 as the basis such that your streaming operator is ½ E1(tau), after all
You can avoid the higher order exponential integrals that arise in Eqn 5-38 onwards by using a stochastic approximation to the integral.

damianjilk2 commented:
NOTE: eq 5-29 from Lewis and Miller is the same as [scipy.special.expn](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.expn.html).
Thus, we can use scipy.special.exp1(tau) in order to solve E1(tau).

damianjilk2 committed:
Updated streaming operator to use scipy exponential integral solver. Revamped positions from input_data.json to be a dictionary that accepts multiple values per voxel. Updated tester file for new datatypes and calculations.

damianjilk2 committed:
Updated streaming operator to use scipy exponential integral solver. Revamped positions from input_data.json to be a dictionary that accepts multiple values per voxel. Updated tester file for new datatypes and calculations.

damianjilk2 committed:
Add scipy to requirements.txt

damianjilk2 committed:
Add scipy to requirements.txt

damianjilk2 committed:
Fixed scipy requirement mistake

damianjilk2 committed:
Fixed scipy requirement mistake

damianjilk2 commented:
This PR allows the user to input a dictionary of positions that they would like to be calculated rather than defaulting to the left edge of the voxel. 

This implementation averages the streaming operators if there are several values within one voxel, ensuring that all combinations are iterated through and calculated. 

A bug currently exists with the equation used for the streaming operator and will be addressed further in a different PR (Issue #36).

The tester script has been modified as well to account for the new datatype used for positions as well as for the new equation used to solve for the streaming operator. 

Please make note of issues with readability or noticeable bugs or problematic code within the scripts. Thank you for reviewing!

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
-    - ri (int): Index of the starting voxel.
-    - rj (int): Index of the ending voxel.
-    - voxel_size (float): Size of each voxel.
+    - start_index (int): Starting position voxel index
+    - end_index (int): Ending position voxel index
```
You should be able to compute/find `start_index` and `end_index` from `ri` and `rj` and the voxel grid.  This will prevent user error.

This is particularly easy for a fixed voxel size, but you may want to plan on dong it for an arbitrary voxel size.

damianjilk2 replied:
Possible solution for arbitrary voxel size: user inputs dictionary similar to
`voxels: {
            "index": 0,
            "start_position": 0.0,
            "end_position": 1.0,
            "cross_section": 0.2,
            "positions": [0.0, 0.1, 0.4]
        },
        {...}`

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+    Example Diagrams:
     
-    TODO: Update this function to accurately calculate tau for two arbitrary points within the voxelized space, considering the actual path of transport.
+      -->  Streaming to the right -->
+    ----------------------------------------------------
+    | Start Voxel  | In-between voxels |  End Voxel  | 
+    ----------------------------------------------------
+    |---[]---------|-------------------|---------[]--|
+    |   ri         |                   |         rj  |
+        |__________|                   |__________|
+         left distance                  right distance
+         
+                  
+      <--  Streaming to the left <--
+    ----------------------------------------------------
+    | End Voxel    | In-between voxels | Start Voxel | 
+    ----------------------------------------------------
+    |---[]---------|-------------------|---------[]--|
+    |   rj         |                   |         ri  |
+        |__________|                   |__________|
+         left distance                  right distance
```
Transport is symmetric so it doesn't actually matter which way you are going.

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
         raise ValueError("Voxel size must be positive and non-zero.")
-    if ri < 0 or ri >= len(sigma_s) or rj < 0 or rj >= len(sigma_s):
-        raise ValueError("Voxel indices must be non-negative and within the range of the sigma_s_matrix.")
+    if ri < 0 or ri > len(sigma_s) * voxel_size or rj < 0 or rj > len(sigma_s) * voxel_size:
+        raise ValueError("Positions must be within the spatial domain defined by the voxels.")
```
Do this before your debugging line above - fail early

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+        logging.debug("Intra-voxel tau: %f", tau)
+    else:
+        # Multiple voxel calculation
+        if ri < rj:
```
Our operator is symmetric, this doesn't matter!

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+            # Direction is to the right (start_index and ri are on the left)
+            
+            # Calculate left fraction
+            if ri == start_index*voxel_size:
+                # ri is on the left-most point of the voxel
+                left_distance = voxel_size
+            elif start_index*voxel_size < ri and ri < (start_index+1)*voxel_size:
+                # ri is between the voxel edges
+                left_distance = (start_index+1)*voxel_size - ri
+            elif ri == (start_index+1)*voxel_size:
+                # ri is on the right-most point of the voxel
+                left_distance = 0
+            else:
+                # ri is not within or on the edges of this voxel
+                raise ValueError(f"Invalid position (ri: {ri}) and voxel index (start_index: {start_index}) combination.")
```
Why isn't this always
```
left_distance = (start_index + 1) * voxel_size - ri
```
no need for all these conditionals, and you'll never have a `ValueError` if you computer `start_index` from `ri`

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+            left_tau = sigma_s[start_index] * left_distance
+            
+            # Calculate right fraction
+            if rj == end_index*voxel_size:
+                # rj is on the left-most point of the voxel
+                right_distance = 0
+            elif end_index*voxel_size < rj and rj < (end_index+1)*voxel_size:
+                # rj is beteween the voxel edges
+                right_distance = rj - (end_index*voxel_size)
+            elif rj == (end_index+1)*voxel_size:
+                # rj is on the right-most point of the voxel
+                right_distance = voxel_size
+            else:
+                # rj is not within or on the edges of this voxel
+                raise ValueError(f"Invalid position (rj: {rj}) and voxel index (end_index: {end_index}) combination.")
```
same as above for the `left_distance`

gonuke commented on drt_1d.py:
```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+                raise ValueError(f"Invalid position (rj: {rj}) and voxel index (end_index: {end_index}) combination.")
+            right_tau = sigma_s[end_index] * right_distance
+            
+            # Sum full voxels
+            center_tau = sum(sigma_s[start_index + 1:end_index]) * voxel_size
+            
+            logging.debug("Direction is to the right. Left_tau: %f, Right_tau: %f, Center_tau: %f", left_tau, right_tau, center_tau)
+            tau = left_tau + right_tau + center_tau
```
Slightly simpler to do:
```suggestion
            path = [left_distance] + [voxel_size] * (end_index - start_index - 1) + [right_distance]
            # Sum full voxels
            tau = np.dot(sigma_s[start_index:end_index+1],  np.array(path))
```

gonuke commented on drt_1d.py:
```diff
@@ -88,10 +195,50 @@ def construct_transition_matrix(num_voxels:int, voxel_size:float, sigma_s:np.nda
     for i in range(num_voxels):
         for j in range(num_voxels):
```
Transition matrix is symmetric, so only need to do lower/upper triangle and copy.

You can also handle the diagonal directly first (without a conditional) and then handle everything up to the diagonal:
```
for j in range(i):

gonuke commented on drt_1d.py:
```diff
@@ -88,10 +195,50 @@ def construct_transition_matrix(num_voxels:int, voxel_size:float, sigma_s:np.nda
+                    num_position_pairs = 0
+                    for position_a, position_b in position_pairs:
+                        # Calculate the streaming operator for the pair (both directions)
+                        streaming_operator_sum += calculate_K_trans(i, j, position_a, position_b, voxel_size, sigma_s)
+                        streaming_operator_sum += calculate_K_trans(i, j, position_b, position_a, voxel_size, sigma_s)
+                        num_position_pairs += 2
```
1. streaming operator is symmetric, don't need to go both ways
2. `num_position_pairs` will always be `2 * len(position_pairs)` so why both accumulating

gonuke commented on drt_1d.py:
```diff
@@ -88,10 +195,50 @@ def construct_transition_matrix(num_voxels:int, voxel_size:float, sigma_s:np.nda
+                positions_in_voxel_j = positions.get(str(j))
+                
+                streaming_operator_sum = 0.0
+                num_position_pairs = 0
+                
+                # Iterate over all combinations of positions between voxel i and voxel j
+                for position_i in positions_in_voxel_i:
+                    for position_j in positions_in_voxel_j:
+                        # Calculate the streaming operator for the pair
+                        streaming_operator_sum += calculate_K_trans(i, j, position_i, position_j, voxel_size, sigma_s)
+                        num_position_pairs += 1
```
`num_position_pairs` will always be `len(positions_in_voxel_i) * len(positions_in_voxel_j)` so just use that number instead of accumulating one at a time

gonuke commented on drt_1d.py:
```diff
@@ -88,10 +195,50 @@ def construct_transition_matrix(num_voxels:int, voxel_size:float, sigma_s:np.nda
             else:
-                K_trans[i, j] = calculate_K_trans(i, j, voxel_size, sigma_s)
+                # Off-diagonal element (different voxels)
+                positions_in_voxel_i = positions.get(str(i))
+                positions_in_voxel_j = positions.get(str(j))
```
Why not:
```suggestion
                positions_in_voxel_i = positions[str(i)]
                positions_in_voxel_j = positions[str(j)]
```

(even better if you convert the keys to integers when you read the data)

gonuke commented on input_data.json:
```diff
@@ -1,15 +1,34 @@
     "maximum_x": 10.0,
-    "number_of_voxels": 7,
+    "positions": {
+        "0": [
```
Can you use integer keys instead of strings?
```suggestion
        0 : [
```

(and then avoid some of my other comments about converting these to integers)

gonuke commented on drt_1d.py:
```diff
@@ -152,14 +299,16 @@ def perform_calculation(input_data:dict):
     sigma_s_values = input_data.get("scattering_cross_section")
-
+    positions = input_data.get("positions")
+    
```
convert keys to integer right away...
```
    positions = { int(key): value for key, value in positions }
```

(or maybe this is not necessary if you use integer keys to begin with - see below)

gonuke commented on drt_1d.py:
```diff
@@ -207,11 +356,11 @@ def setup_logging(level:int = logging.INFO):
 def main():
-    setup_logging()
+    setup_logging(logging.DEBUG)
     logging.debug("Main function started")
```
This is an awful lot of output to be leaving around, even for debugging.

gonuke reviewed (changes requested):
A lot of simplification and streamlining is still possible here.  I think you may be overthinking some of the steps and not relying enough on things that must be true.

damianjilk2 committed:
rebasing with main

damianjilk2 committed:
Created a Voxel object and implemented it into all methods. Updated code based on PR review and improved tester script.

damianjilk2 committed:
Removed copy from requirements.txt. It is a build-in library.

damianjilk2 commented:
Many code changes since last PR review. 
- Created a Voxel object to implement OOP.
- Reduced computation time by only calculating the upper triangle of K_trans since transport is symmetric. 
- Simplified calculate_K_trans significantly by calculating transport in just one direction (because of symmetry) and ensured edge cases still are computed as expected. 
- Created a validate_input method to check the user input.
- Updated input_data.json to include a list of voxel information. Future work will result in removing the scattering cross-section since this will be calculated with ray-tracing. 
- Future work will include allowing voxels to be different sizes. 
- Adjusted number of logging comments. This will be further reduced after more features are included and many voxels are considered in the calculations.

gonuke commented on drt_1d.py:
```diff
@@ -156,32 +252,113 @@ def perform_calculation(input_data:dict):
+    - ValueError: If any validation checks fail.
+    """
+    logging.debug("Validating input data")
+    try:
```
I think an overarching `try` block like this is a big hammer.  You are doing a lot of careful checking and raising errors where they occur - you don't need a try block

gonuke commented on input_data.json:
I'm not convinced by this data structure, but since it's only a toy code, it's probably OK.

Most importantly, it's pretty verbose in the event that one ever wants to generate a large system.  It also has the possibility of having gaps in the geometry, although your validation may take care of that. It's probably better to design cleaner input than to require so much validation.

gonuke commented on drt_1d.py:
```diff
@@ -8,8 +8,33 @@
+import scipy.special as sc
 
-def construct_source_term(num_voxels:int, source_voxel_index:int, source_strength:float=1):
+class Voxel:
```
Not sure this is the best basis for an object in this case...

gonuke commented on drt_1d.py:
```diff
@@ -156,32 +252,113 @@ def perform_calculation(input_data:dict):
-    voxel_size = (x_max - x_min) / num_voxels
+    
+    voxels_data = input_data['voxels']
+    if not isinstance(voxels_data, list) or len(voxels_data) == 0:
```
You already validated the data - it seems redundant to do it again.

gonuke commented on drt_1d.py:
```diff
@@ -145,7 +241,7 @@ def read_input(file_path:str):
     return data
 
-def perform_calculation(input_data:dict):
+def perform_calculation(input_data:dict) -> np.ndarray:
```
I think you can pass in the components of the input data that you need so that this method doesn't need to depend on the structure of `input_data` but instead depends on clearly defined parameters that you can document in the docstring

gonuke commented on drt_1d.py:
```diff
@@ -23,106 +48,177 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
-        
+    # NOTE: Assume sigma_s is still voxel based (will change with continuous geometry)
+    
+    if not voxels:
```
I appreciate that this method could be called from various places, but this is now the 3rd time you've validated this data.  It leads to a lot of bloat.

gonuke reviewed (changes requested):
While I appreciate good defensive programming that validates all your data, I think you could do it more concisely.

The choice of Voxel as the object here may not be serving you well and definitely doesn't lend itself to a readable solution.

damianjilk2 committed:
Removed unnecessary validation causing bloat. Revamped input_data.json to easily handle large numbers of voxels.

damianjilk2 committed:
retrying ci testing

damianjilk2 committed:
Added capability of varying voxel sizes.

gonuke commented on drt_1d.py:
```diff
@@ -145,43 +210,144 @@ def read_input(file_path:str):
+        
+        required_keys = ['name', 'width', 'num_voxels', 'positions_per_voxel',
+                         'position_location', 'scattering_cross_section']
+        for key in required_keys:
+            if key not in region:
+                raise ValueError(f"Region is missing required key: {key}")
```
You can use a set operation to find all the missing keys at once.  Something like this...
```
missing_keys = set(required_keys) - set(region.keys())
if missing_keys:
   # raise error 
```

gonuke commented on drt_1d.py:
```diff
@@ -145,43 +210,144 @@ def read_input(file_path:str):
+        
+        voxel_width = width / num_voxels
+        for voxel_index in range(num_voxels):
+            start_position = start_x + index*width + voxel_index*voxel_width
```
This is incorrect since each region can have a different width.

One approach would be to update `start_x` to be the left side of each region as you loop through the regions

gonuke reviewed (changes requested):
This is definitely a more common way to define your domain at the user input stage.  We should probably discuss the data design.

damianjilk2 committed:
Optimized key validation in validate_input and corrected start_position calculation in perform_calculation

damianjilk2 commented:
Per latest PR review:
- corrected start_position bug within perform_calculation. Now accounts for differences in region widths.
- simplified key validations within validation_input.

gonuke commented:
Some more of these things could be precomputed/pregathered.

For example, you could have
```
# all sigmas
sigmas = np.array([region.sigma for region in regions])

# then use a slice of this in the dot product
tau = np.dot(widths, sigmas[start_idx, end_idx+1])
```

gonuke commented:
There may also be value in another object, say `Domain` with members:
* list of Regions
* boundaries
* widths
* sigmas

as precomputed np.arrays for the data you'll most frequently be accessing

damianjilk2 committed:
New data design per last PR review

damianjilk2 commented:
Since last PR review:

- Data model:
Region and Domain class with all needed methods for computing neutron flux (including source term, optical thickness, construct transition matrix).

- Transition Matrix Construction:
Refactored construct_transition_matrix method to iterate through all voxels and calculate the average streaming operator for all position combinations of two voxels.
Used a dictionary to map voxel IDs to positions within voxels for easy access and iteration.

- Testing
Updated tester script with the new data model.

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
-    s[source_voxel_index] = source_strength
-    logging.debug(f"Constructed source vector: {s}")
-    return s
+    def generate_positions(self, start_x: float) -> np.ndarray:
```
modify this to generate positions in a single voxel given discussion of how we'll use this

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        self.boundaries = self._compute_boundaries()
+        self.widths = np.array([region.width for region in regions])
+        self.sigmas = np.array([region.scattering_cross_section for region in regions])
+        self.num_voxels = sum(region.num_voxels for region in self.regions)
```
It might make sense to store the `cumsum` of the number of voxels, as well, to easily convert a voxel index to which region it's in

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        """
+        return np.cumsum([self.start_x] + [region.width for region in self.regions])
+        
+    def _generate_positions(self) -> np.ndarray:
```
probably don't need this at all

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+            start_x += region.width
+        return np.array(positions)
+
+    def get_region_id(self, x:float) -> int:
```
May want a version of this that takes the voxel index as an input, as well.

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        Returns:
+        - np.ndarray: Array of boundary positions.
+        """
+        return np.cumsum([self.start_x] + [region.width for region in self.regions])
```
Maybe store this in a variable and then update all the regions to know their start_x:
```suggestion
        self.boundaries = np.cumsum([self.start_x] + [region.width for region in self.regions])
        for start_x, region in zip(self.boundaries[:-1],self.regions):
            region.start_x = start_x
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        """
+        self.start_x = start_x
+        self.regions = regions
+        self.boundaries = self._compute_boundaries()
```
Given changes below
```suggestion
        self._compute_boundaries()
```

gonuke replied:
Move this to after where you set all the other variables

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        region_idx = self.get_region_id(x)
+        region = self.regions[region_idx]
+        local_x = x - self.boundaries[region_idx]
+        return int(np.floor(local_x / region.voxel_width) + np.sum([reg.num_voxels for reg in self.regions[:region_idx]]))
```
this would use the cumsum of the voxel counts mentioned above

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
-    """
-    Calculate the neutron flux vector.
+        # Create a dictionary where the keys are voxel ids and values are list of positions within that voxel.
+        voxel_positions = dict()
+        for pos in self.positions:
+            voxel_id = self.get_voxel_id(pos)
+            if voxel_id not in voxel_positions:
+                voxel_positions[voxel_id] = []
+            voxel_positions[voxel_id].append(pos)
```
remove this in favor or generating/querying positions on the fly as you move through the voxels

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+            voxel_positions[voxel_id].append(pos)
+        
+        # Now iterate through all of the combinations of voxel pairs and average the position combinations within those voxels.
+        for i in range(self.num_voxels):
```
track the region ID as you move through the voxels to avoid having to look it up/search

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+                        streaming_sum += self.compute_streaming_operator(pos_i, pos_j)
+                avg_streaming_operator = streaming_sum / (num_positions_i * num_positions_j)
+                K_trans[i,j] = avg_streaming_operator
+                K_trans[j,i] = avg_streaming_operator
```
```suggestion
                K_trans[j,i] = K[i,j]
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,276 @@
+        K_trans = self.construct_transition_matrix()
+        logging.info(f"Constructed transition matrix: {K_trans}")
+        
+        sigma_s = []
+        for region in self.regions:
+            sigma_s.extend([region.scattering_cross_section] * region.num_voxels)
+        sigma_s = np.array(sigma_s)
```
```suggestion
        sigma_s = np.zeros(total_num_voxels)
        for region in self.regions:
            sigma_s[region.first_voxel_idx:region.last_voxel_idx] = region.scattering_cross_section
```

gonuke reviewed (changes requested):
Some modest changes discussed during our meeting

damianjilk2 committed:
PR review update

damianjilk2 commented:
Since last PR review:

- Improved `construct_transition_matrix` by incrementing region index, utilizing pre-computed `cumsum_voxels` value, generating positions on the fly, and passing region indices to `compute_streaming_operator` and `compute_optical_thickness`.
- Updated position generation in `Region` class to be done on a voxel basis rather than region basis. 
- Removed position generation in `Domain` class.
- Refactored `sigma_s` matrix to use `cumsum_voxels` and initialize using np.array rather than a list.

damianjilk2 committed:
Update ci.yml due to deprecated versions

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+        self.regions = regions
+        self.widths = np.array([region.width for region in regions])
+        self.sigmas = np.array([region.scattering_cross_section for region in regions])
+        self.num_voxels = sum(region.num_voxels for region in self.regions)
+        self.cumsum_voxels = np.cumsum([0] + [region.num_voxels for region in self.regions])
```
Swap the order these are defined and use one to define the other
```suggestion
        self.cumsum_voxels = np.cumsum([0] + [region.num_voxels for region in self.regions])
        self.num_voxels = self.cumsum_voxels[-1]
```

But could also define `num_voxels` as a property to just return this
```
    @property
    def num_voxels(self):
        return self.cumsum_voxels[-1]
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+        self.cumsum_voxels = np.cumsum([0] + [region.num_voxels for region in self.regions])
+        self._compute_boundaries()
+        
+    def _compute_boundaries(self) -> np.ndarray:
```
Since this is only ever called by `_init_()` and it's only a couple of lines, you might as well put it in `_init_()`

This may seem contrary to other discussion about modularity and breaking things into functions, but it's one of those gray areas.  In this case, your `_init_()` function is relatively streamlined, and this is definitely part of initialization.

Alternatively, you could add a method called by `_init_()` that processes all the `Region` data into `Domain` variables: `widths`, `sigmas`, `cumsum_voxels`, `boundaries`
And then also sets the `Region` data: `start_x` and maybe `first_voxel_id` (see below)

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+        """
+        region_idx = self.get_region_id_from_position(x)
+        region = self.regions[region_idx]
+        local_x = x - self.boundaries[region_idx]
+        voxel_id_within_region = int(np.floor(local_x / region.voxel_width))
```
I think there is value in making this a method of the `Region` class:
```
    def get_local_voxel_id(self, x):
        return int(np.floor( (x - self.start_x) / region.voxel_width))
```

Bonus: if you store the first voxel id for each region in the same way you store the start x for each region, at initialization, you could include that here too  

```
    def get_global_voxel_id(self, x):
        return self.get_local_voxel_id(x) + self.first_voxel_id
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+        - float: Width of the voxel containing the x-coordinate.
+        """
+        region_idx = self.get_region_id_from_position(x)
+        region = self.regions[region_idx]
+        return region.voxel_width
```
Here is a place to be more concise
```suggestion
        return self.regions[region_idx].voxel_width
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+        
+        for i in range(self.num_voxels):
+            # Update region index if necessary.
+            while i >= self.cumsum_voxels[region_idx_i + 1]:
```
This will happen in an orderly and incremental fashion, I think a while loop creates confusion about that
```suggestion
            if i >= self.cumsum_voxels[region_idx_i + 1]:
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+            # Update region index if necessary.
+            while i >= self.cumsum_voxels[region_idx_i + 1]:
+                region_idx_i += 1
+            region_i = self.regions[region_idx_i]
```
Initialize this before the loop and then only change this when `region_idx_i` changes.

This is a readability/cognitive load issue - it is useful to group things that are only relevant under certain circumstances.

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+            region_idx_j = region_idx_i
+            
+            for j in range(i, self.num_voxels):
+                while j >= self.cumsum_voxels[region_idx_j + 1]:
+                    region_idx_j += 1
+                
+                region_j = self.regions[region_idx_j]
```
as above
```suggestion
                if j >= self.cumsum_voxels[region_idx_j + 1]:
                    region_idx_j += 1
                    region_j = self.regions[region_idx_j]
```

gonuke commented on drt_1d.py:
```diff
@@ -8,121 +8,289 @@
+            region_i = self.regions[region_idx_i]
+            
+            # Find start_x for voxel_i = region_edge + num_voxels_incremented*voxel_width.
+            voxel_start_x_i = self.boundaries[region_idx_i] + (i - self.cumsum_voxels[region_idx_i]) * region_i.voxel_width
```
If you store `first_voxel_id` in a region (referenced elsewhere) then
```suggestion
            voxel_start_x_i = region_i.start_x + (i - region_i.first_voxel_id) * region_i.voxel_width
```

gonuke replied:
Could consider making this a method in `Region` since it uses so much data from the region

gonuke replied:
Thinking more... you only need this `voxel_start_x` position to use in `generate_positions` which is a method in `Region`.  So what if `generate_positions` took the voxel ID as an argument instead of the voxel starting position, and you calculate that within `generate_positions`?

gonuke reviewed (changes requested):
A few more comments here... we're really at a polishing stage here, but I see opportunities for learning some of the finer details. However, may be we need to not get too far down the polishing before you master some of the main design ideas.

damianjilk2 committed:
PR polishing revision changes

damianjilk2 commented:
Changes since last review:
- first_voxel_id value added to Region class.
- get_local_voxel_id and get_local_voxel_id methods added to Region class.
- generate_positions in the Region class now takes the voxel_id as an input rather than x_start. 
- _compute_boundares was removed from the Domain class and added to the __init__ instead. 
- In addition to updating start_x values for each region, the __init__ of the Domain class assigns the first_voxel_id value.
- num_voxels within the Domain class changed to be a property which utilizes cumsum_voxels.
- get_voxel_id method removed from Domain class since there is now a local and global version of this at the region level. 
- readability changes to construct_transition_matrix.

gonuke commented:
This looks good now.  We're definitely getting to the asymptotic of perfect...
